pkg/cloud/baremetal/actuators/machine/actuator.go

```diff
@@ -430,6 +438,165 @@ func consumerRefMatches(consumer *corev1.ObjectReference, machine *machinev1.Mac
 	return true
 }
 
+// updateUserData fetches full ignition from MCS and updates the UserData secret reference
+func (a *Actuator) updateUserData(ctx context.Context, config *bmv1alpha1.BareMetalMachineProviderSpec,
+	machine *machinev1.Machine) error {
+
+	// Try to fetch the full ignition secret to see if it exists
+	machineUserDataSecret := &corev1.Secret{}
+	key := client.ObjectKey{
+		Name:      machine.Name + "-user-data",
+		Namespace: machine.Namespace,
+	}
+	// Secret already exists. Set secretReference and return
+	err := a.client.Get(ctx, key, machineUserDataSecret)
+	if err == nil {
+		// Set Machine's secretreference
+		config.UserData = &corev1.SecretReference{
+			Name:      machine.Name + "-user-data",
+			Namespace: machine.Namespace,
+		}
+		return nil
+	}
+
+	// userDataSecret is the pointer ignition originally
+	// created by openshift installer. Fetch it to extract
+	// its content (an URL containg full ignition and a
+	// CA cert for the URL's https connection)
+	userDataSecret := corev1.Secret{}
+	key = client.ObjectKey{
+		Name:      config.UserData.Name,
+		Namespace: config.UserData.Namespace,
+	}
+	err = a.client.Get(ctx, key, &userDataSecret)
+	if err != nil {
+		log.Printf("Cannot get %s secret: %s", config.UserData.Name, err.Error())
+		return err
+	}
+	encodedUserDataBytes := userDataSecret.Data["userData"]
+	var userData []byte
+	userData, err = base64.StdEncoding.DecodeString(string(encodedUserDataBytes))
+	if err != nil {
+		log.Printf("Error base64 decoding userData : %s", err.Error())
+		return fmt.Errorf("error base64 decoding userData: %w", err)
+	}
+
+	// Define an ignition data struct to unmarshal incoming json
+	type IgnitionData struct {
+		Ignition struct {
+			Config struct {
+				Append []struct {
+					Source string
+				}
+			}
+			Security struct {
+				TLS struct {
+					CertificateAuthorities []struct {
+						Source string
+					}
+				}
+			}
+		}
+	}
+	var ignData IgnitionData
+
+	if err := json.Unmarshal(userData, &ignData); err != nil {
+		log.Printf(string(userData))
+		log.Printf("Cannot unmarshal ignition config from userData : %s", err.Error())
+		return fmt.Errorf("could not unmarshal ignition config from userData: %w", err)
+	}
+	// Fetch IgnitionURL and the CA cert
+	ignitionURL := ignData.Ignition.Config.Append[0].Source
+	caCertRaw := ignData.Ignition.Security.TLS.CertificateAuthorities[0].Source
```

Comment:
Is it possible to successfully parse the JSON and still have empty Append or CertificateAuthorities arrays?

Reply:
In the case of empty Append and CertificateAuthorities, the json will not unmarshal correctly and we'll fail before reaching this point.

Reply:
It's *definitely* possible to panic here, should always validate the length of an unknown slice before accessing a member -- here's an example of a panic using this code:

https://play.golang.org/p/jEB7sSoAAVa

Reply:
Ha!! Good catch. My own test code had a malformed json and I convinced myself that the json unmarshal will catch it. Thanks for pointing this out. Fix incoming.